src/main/java/com/iota/iri/service/milestone/MilestoneSolidifier.java

```diff
@@ -0,0 +1,247 @@
+
+        earliestMilestoneHash = nextEarliestMilestone.getKey();
+        earliestMilestoneIndex = nextEarliestMilestone.getValue();
+        earliestMilestoneTries = 0;
+    }
+
+    /**
+     * This method performs the actual solidity check on the selected milestone.
+     *
+     * It first checks if there is a milestone that has to be solidified and then issues the corresponding solidity
+     * check. In addition to issuing the solidity checks, it dumps a log message to keep the node operator informed
+     * about the progress of solidification.
+     *
+     * We limit the amount of transactions that may be processed during the solidity check, since we want to solidify
+     * from the oldest milestone to the newest one and not "block" the solidification with a very recent milestone that
+     * needs to traverse huge chunks of the tangle. If we fail to solidify a milestone for a certain amount of tries, we
+     * increase the amount of transactions that may be processed by using an exponential binary backoff strategy. The
+     * main goal of this is to give the solidification enough "resources" to discover the previous milestone (if it ever
+     * gets stuck because of a very long path to the previous milestone) while at the same time allowing fast solidity
+     * checks in "normal conditions".
+     *
+     * @return true if there are no unsolid milestones that have to be processed or if the earliest milestone is solid
+     */
+    private boolean earliestMilestoneIsSolid() {
+        if (earliestMilestoneHash == null) {
+            return true;
+        }
+
+        if (unsolidMilestones.size() > 1) {
+            statusLogger.status("Solidifying milestone #" + earliestMilestoneIndex + " [" + unsolidMilestones.size() + " left]");
+        }
+
+        try {
+            return transactionValidator.checkSolidity(
+                earliestMilestoneHash,
+                true,
+                2 ^ Math.min(
+                    earliestMilestoneTries++ / SOLIDIFICATION_TRANSACTIONS_LIMIT_INCREMENT_INTERVAL,
+                    SOLIDIFICATION_TRANSACTIONS_LIMIT_MAX_INCREMENT
+                ) * SOLIDIFICATION_TRANSACTIONS_LIMIT
+            );
+        } catch (Exception e) {
+            statusLogger.error("Error while solidifying milestone #" + earliestMilestoneIndex, e);
+
+            return false;
+        }
+    }
+
+    /**
+     * This method checks if the current unsolid milestone has become solid or irrelevant and advances to the next one
+     * if that is the case.
+     *
+     * It is getting called by the solidification thread in regular intervals.
+     *
+     * @return true if the current solidification task was successful and the next milestone is due or false otherwise
+     */
+    private boolean processSolidificationTask() {
+        if (earliestMilestoneHash != null && (
+            earliestMilestoneIndex <= snapshotManager.getInitialSnapshot().getIndex() ||
```

Comment:
How could this ever happen if you only add to the `unsolidMilestones` when:

```
   if (milestoneIndex > snapshotManager.getInitialSnapshot().getIndex()) {
            if (milestoneIndex < earliestMilestoneIndex) {
                earliestMilestoneHash = milestoneHash;
                earliestMilestoneIndex = milestoneIndex;
                earliestMilestoneTries = 0;
            }

            unsolidMilestones.put(milestoneHash, milestoneIndex);
        }
```

Reply:
Since we have 3 threads that solidify transaction, a milestone could become solid outside of this thread. If this happens and a snapshot gets taken it could start pruning the exact same milestone which would make this thread start to solidify it again (and with it the whole tangle) that just started to get pruned.

While this is a very rare edge case it could theoretically happen (and i think i saw it happening once) so i only try to solidify milestones that are still ahead of the cut off point of the last local snapshot (just to be safe).

Note: Traversing the graph can take a really long time if the missing tx is "behind" a huge chunk of known transactions. This will however be solved once we add the bottom->top solidification anyway. But for now its required.